tests/unit/lms/services/grouping_test.py

```diff
@@ -166,15 +171,20 @@ def test_you_can_have_a_group_and_a_section_with_the_same_id(
         )
 
     @pytest.fixture
-    def upsert_with_parent(self, svc):
-        return partial(
-            svc.upsert_with_parent,
-            tool_consumer_instance_guid="t_c_i_guid",
-            lms_id="lms_id",
-            lms_name="lms_name",
-            parent=factories.Course(application_instance=svc.application_instance),
-            type_=Grouping.Type.CANVAS_GROUP,
-        )
+    def upsert_with_parent(self, svc, db_session):
```

Comment:
Maybe the fixutre itself should be named `upsert_with_parent_single_item`?

Reply:
I think we may need to re-design the tests a little now that `upsert_with_parent()` accepts a list of dicts:

* I think we need at least one test that passes it a list of more than one dict

* `upsert_with_parent()` can now insert some new groupings and update some existing groupings at the same time. I think we need a test that does that. We probably want to combine `test_if_no_grouping_already_exists_it_inserts_a_new_one()` and `test_if_a_grouping_already_exists_it_updates_it()` into a single test that both inserts a new grouping and updates an existing one in one call

* I'm not sure that the partial caller fixture for `upsert_with_parent()` makes much sense anymore. That fixture made sense when `upsert_with_parent()` took several separate arguments because the fixture was simple and it could inject defaults for all the arguments. But now that `upsert_with_parent()` takes a list of dicts I think the fixture might just be getting in the way

I had a go at trying to write a test for inserting one row and updating another at the same time. I can't for the life of me get this test to actually work. But here's where I got to:

```python
class TestUpsertWithParent:
    def test_it(self, db_session, grouping_dict, svc):
        course = factories.Course(application_instance=svc.application_instance)

        # Insert an existing grouping into the DB.
        existing_grouping = factories.CanvasSection(parent=course)

        # Upsert one grouping that already exists and one that doesn't exist yet.
        groupings = svc.upsert_with_parent(
            [
                {
                    "lms_id": existing_grouping.lms_id,
                    "lms_name": "updated_name",
                    "parent": existing_grouping.parent,
                    "type": existing_grouping.type,
                    "extra": {"updated": "extra"},
                },
                {
                    "lms_id": "new_grouping_id",
                    "lms_name": "new_grouping_name",
                    "parent": course,
                    "type": Grouping.Type.CANVAS_GROUP,
                    "extra": {"new": "extra"},
                },
            ]
        )

        # It should have updated the existing grouping.
        existing_grouping = db_session.query(Grouping).filter_by(lms_id=existing_grouping.lms_id).one()
        assert existing_grouping.lms_name == "updated_name"
        assert existing_grouping.extra == {"updated": "extra"}

        # It should have created the new grouping.
        new_grouping = db_session.query(Grouping).filter_by(lms_id="new_grouping_id").one()
        assert new_grouping.lms_name == "new_grouping_name"
        assert new_grouping.parent == course
        assert new_grouping.type == Grouping.Type.CANVAS_GROUP
        assert new_grouping.extra == {"updated": "extra"}
```

Reply:
Here's another attempt that avoids using the ORM to insert the `existing_grouping`. This gets past the constraint error but it seems that `upsert_with_parent()` doesn't actually update `existing_grouping`:

```python
    def test_it(self, db_session, grouping_dict, svc):
        course = factories.Course(application_instance=svc.application_instance)
        # Insert an existing grouping into the DB.
        svc.upsert_with_parent(
            [
                {
                    "lms_id": "existing_id",
                    "lms_name": "existing_name",
                    "parent": course,
                    "type": Grouping.Type.CANVAS_SECTION,
                    "extra": {"existing": "extra"},
                }
            ]
        )

        # Upsert one grouping that already exists and one that doesn't exist yet.
        groupings = svc.upsert_with_parent(
            [
                {
                    "lms_id": "existing_id",
                    "lms_name": "updated_name",
                    "parent": course,
                    "type": Grouping.Type.CANVAS_SECTION,
                    "extra": {"updated": "extra"},
                },
                {
                    "lms_id": "new_grouping_id",
                    "lms_name": "new_grouping_name",
                    "parent": course,
                    "type": Grouping.Type.CANVAS_GROUP,
                    "extra": {"new": "extra"},
                },
            ]
        )

        # It should have updated the existing grouping.
        existing_grouping = (
            db_session.query(Grouping).filter_by(lms_id="existing_id").one()
        )
        assert existing_grouping.lms_name == "updated_name"
        assert existing_grouping.extra == {"updated": "extra"}

        # It should have created the new grouping.
        new_grouping = (
            db_session.query(Grouping).filter_by(lms_id="new_grouping_id").one()
        )
        assert new_grouping.lms_name == "new_grouping_name"
        assert new_grouping.parent == course
        assert new_grouping.type == Grouping.Type.CANVAS_GROUP
        assert new_grouping.extra == {"updated": "extra"}

```